### PR TITLE
[System.Private.CoreLib] Methods with no instance access can be made static

### PR DESCRIPTION
**Pull request by marek-safar**

### CODE REVIEW

File: src/coreclr/src/System.Private.CoreLib/src/System/Reflection/Emit/MethodBuilder.cs [hunk 1]
```diff
@@ -135,14 +135,14 @@ public sealed class MethodBuilder : MethodInfo
 
         #region Internal Members
 
-        internal void CheckContext(params Type[]?[]? typess)
+        internal static void CheckContext(params Type[]?[]? typess)
```
Comment (stephentoub) [inline src/coreclr/src/System.Private.CoreLib/src/System/Reflection/Emit/MethodBuilder.cs:138]: Nit: can we fix the argument name as long as we're tweaking this?

File: src/coreclr/src/System.Private.CoreLib/src/System/Reflection/Emit/ModuleBuilder.cs [hunk 2]
```diff
@@ -122,12 +122,12 @@ internal void CheckTypeNameConflict(string strTypeName, Type? enclosingType)
             return SymbolType.FormCompoundType(strFormat, baseType, 0);
         }
 
-        internal void CheckContext(params Type[]?[]? typess)
+        internal static void CheckContext(params Type[]?[]? typess)
```
Comment (stephentoub) [inline src/coreclr/src/System.Private.CoreLib/src/System/Reflection/Emit/ModuleBuilder.cs:125]: Ditto

File: src/coreclr/src/System.Private.CoreLib/src/System/Reflection/Emit/MethodBuilder.cs [hunk 3]
```diff
@@ -135,14 +135,14 @@ public sealed class MethodBuilder : MethodInfo
...
         {
-            m_module.CheckContext(typess);
+            ModuleBuilder.CheckContext(typess);
         }
```
Comment (stephentoub) [inline src/coreclr/src/System.Private.CoreLib/src/System/Reflection/Emit/MethodBuilder.cs:141]: Does this method provide any benefit at this point?  i.e. can callers just use ModuleBuilder.CheckContext directly? (Same goes for the other cases of this below.)

File: src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs [hunk 4]
```diff
@@ -492,13 +492,12 @@ private void RunReadWriteTaskWhenReady(Task asyncWaiter, ReadWriteTask readWrite
                 {
                     Debug.Assert(t.IsCompletedSuccessfully, "The semaphore wait should always complete successfully.");
                     var rwt = (ReadWriteTask)state!;
-                    Debug.Assert(rwt._stream != null);
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:495]: Why did you delete the assert?

Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:495]: I thought it was there because of the instance call (it does not seem to be used inside RunReadWriteTask)

File: src/coreclr/src/System.Private.CoreLib/src/System/Reflection/Emit/MethodBuilder.cs (hunk 1 above)
Comment (marek-safar) [reply src/coreclr/src/System.Private.CoreLib/src/System/Reflection/Emit/MethodBuilder.cs]: Cleaned up

File: src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs (hunk 4 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:495]: There are some confusing code paths here as part of optimizations, but basically RunReadWriteTask invokes some code that will end up nulling out _stream, so this assert is helping to validate that this code isn't run a second time (if it was, it'd be null).

Please put back the assert :)

**Review by stephentoub — APPROVED** [review]
Other than the one comment about putting back the deleted assert, LGTM. Thanks.

How were these found?  Some analyzer?  Is there a lot more to do to get to a clean state where it could be enabled by default?

**Comment by marek-safar** [comment]
@stephentoub found one from customer report and then run the repo analyzer to see how bad things are. There are many more candidates across libraries but we'd need to also suppress the warning public APIs which should be static but we cannot make them as it'd be breaking change.

File: src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs (hunk 4 above)
Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs:495]: Done (with a comment)

**Comment by stephentoub** [comment]
> There are many more candidates across libraries but we'd need to also suppress the warning public APIs which should be static but we cannot make them as it'd be breaking change.

Which analyzer is it?

**Comment by marek-safar** [comment]
> Which analyzer is it?

https://github.com/dotnet/runtime/blob/master/eng/CodeAnalysis.ruleset#L102

**Comment by stephentoub** [comment]
In that case we can add:
```
dotnet_code_quality.ca1822.api_surface = private, internal
```
to the .editor config and it won't fire for public surface area.
